File: mobile-engage/src/main/java/com/emarsys/mobileengage/notification/command/AppEventCommand.java
package com.emarsys.mobileengage.notification.command;

import android.content.Context;

import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.api.event.EventHandler;
import com.emarsys.mobileengage.event.EventHandlerProvider;

import org.json.JSONObject;

public class AppEventCommand implements Runnable {

    private String name;
    private JSONObject payload;
    private Context context;
    private EventHandlerProvider eventHandlerProvider;

    public AppEventCommand(Context context, EventHandlerProvider eventHandlerProvider, String name, JSONObject payload) {
        Assert.notNull(context, "Context must not be null!");
        Assert.notNull(name, "Name must not be null!");
        Assert.notNull(eventHandlerProvider, "EventHandlerProvider must not be null!");
        this.context = context;
        this.eventHandlerProvider = eventHandlerProvider;
        this.name = name;
        this.payload = payload;
    }

    public Context getContext() {
        return context;
    }

    public String getName() {
        return name;
    }

    public JSONObject getPayload() {
        return payload;
    }

    public EventHandler getNotificationEventHandler() {
        return eventHandlerProvider.getEventHandler();
    }

    @Override
    public void run() {
        EventHandler eventHandler = eventHandlerProvider.getEventHandler();
        if (eventHandler != null) {
            eventHandler.handleEvent(context, name, payload);
        }
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/notification/command/AppEventCommand.kt
package com.emarsys.mobileengage.notification.command

import android.content.Context
import com.emarsys.core.Mockable
import com.emarsys.core.handler.ConcurrentHandlerHolder
import com.emarsys.mobileengage.api.event.EventHandler
import com.emarsys.mobileengage.event.CacheableEventHandler
import kotlinx.coroutines.launch
import org.json.JSONObject

@Mockable
class AppEventCommand(
    private val context: Context,
    private val cacheableEventHandler: CacheableEventHandler,
    private val concurrentHandlerHolder: ConcurrentHandlerHolder,
    val name: String,
    val payload: JSONObject?
) : Runnable {

    val notificationEventHandler: EventHandler?
        get() = cacheableEventHandler

    override fun run() {
        val eventHandler = cacheableEventHandler
        concurrentHandlerHolder.uiScope.launch {
            eventHandler.handleEvent(context, name, payload)
        }
    }
}